### PR TITLE
Match git describe using format

### PR DESCRIPTION
**Pull request by kworth**
To take the concept of "match git describe" one step further, this PR makes some further improvements on top of #73 by making the functionality available via two different options:

- Set matchGitDescribe to true 
    - A) Include the letter 'g' in the versionName (resolves issue #66)
    - B) Use the rev count calculated by git (resolves issue #67)
- Insert the string `"%describe%"` anywhere within their `format` (resolves issues called out in #68)

The first option should allow for ease of use so that users expecting the output to match `git describe` (but with no need or desire to modify their format) can simply configure the plugin with a simple boolean toggle.

The second option should allow for flexibility so that anyone who ***does*** want to modify their format has the ability to do so while still matching `git describe`

Note: There is technically no need to set `matchGitDescribe` to true once `"%describe%"` has been added into `format`. This may be confusing to some as they may wonder whether it makes a difference or not. Therefore, we may decide to somehow evolve this PR further to clear up any possible confusion, or it could simply be left for an additional change later down the road.

### CODE REVIEW

**Comment by kworth** [comment]
I think I now prefer this PR over #73 so much so that I will opt to close/reject #73.

**Comment by gladed** [comment]
Can we remove `matchGitDescribe` and just support `%describe%` in this PR? Should be pretty easy and I want to give you credit for the commit. (I'll add README updates, especially a warning about differences between `%describe%` and the built-in version counter.)

**Comment by kworth** [comment]
I haven't quite given up hope of convincing you that `git describe` is the best default. :D Both because it will allow you to clean up the code (by getting rid of custom code which duplicates what the developers of Git have already designed and tested) and because it's the best commit-ish that users of the plugin should want to get out of it.

As an example, I just found the other day that you could (I think) use the `--first-parent` argument in `git describe` to get the rev count you have been preferring. So, perhaps I just need to try to put that into this PR and see if we can't come together on this.

**Comment by gladed** [comment]
Note, https://download.eclipse.org/jgit/site/5.6.0.201912101111-r/apidocs/org/eclipse/jgit/api/DescribeCommand.html does not have this option.

**Comment by kworth** [comment]
Well that's lame. 😣

I'm not concerned about credit, so if there's a desire to do it differently than with these commits, I won't lose any sleep over this code being rejected.

Mainly I want the functionality to be supported and hopefully in such a way that will make this plugin the go-to popular way to quickly/easily set versionName and versionCode (and additionally allow for people to modify and customize it...). Currently - from my point of view - the plugin is not built for the most likely/common usage, out of the box. So it would feel odd for me to recommend this to people and then have to tell them they need to configure it to get what (in my mind) is the most likely desired output.

Again, I fully recognize this as a matter of opinion and thus I take no hard stance on trying to insist one way or the other - especially when I'm just recently inserting myself into this.

Thanks for your time.

**Comment by gladed** [comment]
Oh I would definitely want to pick some defaults differently. Going to 1.0 might make this possible (Major version). I've opened some other issues to cover these.

Making `--first-parent` the default behavior is another good idea. This would prevent the merge-from-release branch problem (see #22), using `tagPattern` for this is a hack. I've opened #75 for this.

Before going to 1.0 I would like solution which affords control over "commits since last tag" (#67). Sometimes, you would want it minimized, and in other cases you might want it to match git-describe. I'm with you that matching git-describe as a default is probably best.

I think our only difference is in implementation strategy. After investigation, we know:
* `jgit`'s describe function is not sufficient: it can't support other use cases here.
* `jgit`'s code for this is only a page or so: re-implementing shouldn't be onerous.
* `jgit`'s code isn't the canonical implementation: command-line `git` is the flagship, and there may well be cases where these two tools don't align 100% anyway.

My conclusion from all of that is to say: yup, we should re-implement here. It's just a matter of doing the work :)

Coming back to this pull, in a 0.x branch I'm fine with offering `%describe%` as a quick fix to cover your use case. It probably won't be necessary in 1.0 since we can make the default behavior identical to `git describe --first-parent --tags` (including prepending a `g` to the commit tag because, git!) (dang need another issue for that).

I hope this sounds fair? I really appreciate your thoughts.

**Comment by kworth** [comment]
Actually, on the last point, as soon as you use `git describe` then it will put the `'g'` on there itself, so you shouldn't need to do anything extra.

**Comment by kworth** [comment]
@gladed this should be good to go now, I think.

File: src/main/groovy/com/gladed/androidgitversion/AndroidGitVersion.groovy [hunk 1]
```diff
@@ -41,6 +41,7 @@ class AndroidGitVersion implements Plugin<Project> {
 }
 
 class AndroidGitVersionExtension {
+
```
Comment (gladed) [inline src/main/groovy/com/gladed/androidgitversion/AndroidGitVersion.groovy:44]: Can we remove the extra blanks?

File: src/main/groovy/com/gladed/androidgitversion/AndroidGitVersion.groovy [hunk 2]
```diff
@@ -397,6 +406,9 @@ class AndroidGitVersionExtension {
         /** Most recent version seen */
         String lastVersion = 'unknown'
 
+        /** The resulting output from calling git describe */
+        String outputOfGitDescribe = null
```
Comment (gladed) [inline src/main/groovy/com/gladed/androidgitversion/AndroidGitVersion.groovy:410]: How about `gitDescribeOutput`.

Can this default to `unknown`? Then we don't need to test for null @ line 140.

File: src/test/groovy/com/gladed/androidgitversion/FormatTest.groovy [hunk 3]
```diff
@@ -30,6 +30,40 @@ class FormatTest extends AndroidGitVersionTest {
         assert plugin.name().endsWith(').release_1.x')
     }
 
+    void testBranchNameOnGitDescribe() {
+        addCommit()
+        addTag("1.0")
+        addBranch("feature-foo")
+        new File(projectFolder.root, "build.gradle").append("// addition 1")
```
Comment (gladed) [inline src/test/groovy/com/gladed/androidgitversion/FormatTest.groovy:37]: Don't need this line, `addCommit()` does this already

File: src/test/groovy/com/gladed/androidgitversion/FormatTest.groovy [hunk 4]
```diff
@@ -30,6 +30,40 @@ class FormatTest extends AndroidGitVersionTest {
...
+        plugin.format = "%describe%%-branch%%-dirty%"
+
+        assert plugin.name().equals("1.0-dirty")
+    }
+
+    void testDirtyBranchOnGitDescribe() {
+        addCommit()
+        addTag("1.0")
+        addBranch("release/1.x")
+        new File(projectFolder.root, "build.gradle").append("// addition 1")
```
Comment (gladed) [inline src/test/groovy/com/gladed/androidgitversion/FormatTest.groovy:58]: Can remove

File: src/test/groovy/com/gladed/androidgitversion/MainTest.groovy [hunk 5]
```diff
@@ -160,6 +161,34 @@ class MainTest extends AndroidGitVersionTest {
         assertTrue("untracked is dirty", plugin.name().contains("dirty"))
     }
 
+    void testMatchGitDescribeOffByDefault() {
```
Comment (gladed) [inline src/test/groovy/com/gladed/androidgitversion/MainTest.groovy:164]: I think we're covering both these cases in FormatTest, can we remove?

**Review by ghost — COMMENTED** [review]
Looks good, just a couple housekeeping items

**Comment by gladed** [comment]
Closing, superseded by #78
